src/main/java/org/truffleruby/debug/TruffleDebugNodes.java

```
@@ -133,33 +137,53 @@ public DynamicObject printBacktrace() {
-        public DynamicObject astMethod(DynamicObject method) {
-            return ast(Layouts.METHOD.getMethod(method));
+        public DynamicObject astMethod(DynamicObject method, NotProvided block) {
+            ast(Layouts.METHOD.getMethod(method));
+            return nil();
         }
 
+        @TruffleBoundary
         @Specialization(guards = "isRubyUnboundMethod(method)")
-        public DynamicObject astUnboundMethod(DynamicObject method) {
-            return ast(Layouts.UNBOUND_METHOD.getMethod(method));
+        public DynamicObject astUnboundMethod(DynamicObject method, NotProvided block) {
+            ast(Layouts.UNBOUND_METHOD.getMethod(method));
+            return nil();
         }
 
+        @TruffleBoundary
         @Specialization(guards = "isRubyProc(proc)")
-        public DynamicObject astProc(DynamicObject proc) {
-            return ast(Layouts.PROC.getMethod(proc));
+        public DynamicObject astProc(DynamicObject proc, NotProvided block) {
+            ast(Layouts.PROC.getCallTargetForType(proc));
+            return nil();
         }
 
         @TruffleBoundary
+        @Specialization
+        public DynamicObject astBlock(NotProvided proc, DynamicObject block) {
+            ast(Layouts.PROC.getCallTargetForType(block));
+            return nil();
+        }
+
         private DynamicObject ast(InternalMethod method) {
-            if (method.getCallTarget() instanceof RootCallTarget) {
-                return ast(((RootCallTarget) method.getCallTarget()).getRootNode());
+            return ast(method.getCallTarget());
+        }
+
+        private DynamicObject ast(CallTarget callTarget) {
+            if (callTarget instanceof RootCallTarget) {
+                return ast((RootCallTarget) callTarget);
             } else {
-                return nil();
+                throw new RaiseException(getContext().getCoreExceptions().internalError("call target is not a root call target", this));
```

I think that's a recipe for confusing behaviour in the future. The previous code was just silently doing nothing in some unexpected cases!